### PR TITLE
Filter Short Baselines, DD matrix and Sum in Complex Domain 

### PR DESCRIPTION
**Pull request by calgray**
This feature adds a command-line option to specify a minimum baseline length to filter with in phase matrix calibration. This also add a small log at calibration start for how many baselines were filtered.

### CODE REVIEW

File: src/icrar/leap-accelerate/algorithm/casa/PhaseRotate.cc [hunk 1]
```diff
@@ -154,53 +153,48 @@ namespace casalib
...
 
                 // Calculate DInt
                 casacore::Matrix<double> dInt = casacore::Matrix<double>(metadata.I.size() + 1, phaseAngles.shape()[1]);
                 dInt = 0;
+                constexpr double two_pi = boost::math::constants::pi<double>();
                 for(size_t n = 0; n < metadata.I.size(); ++n)
                 {
-                    dInt.row(n) = phaseAnglesI.row(n) - (casacore::sum((casacore::Array<double>)metadata.A.row(n) * (casacore::Array<double>)cal1.column(0)));
+                    double sum = casacore::sum((casacore::Array<double>)metadata.A.row(n) * (casacore::Array<double>)cal1.column(0));
+                    std::complex<double> scalear = std::exp(-sum*two_pi) + 0i;
```
Comment (rtobar) [inline src/icrar/leap-accelerate/algorithm/casa/PhaseRotate.cc:190]: `std::complex<double> scalar {std::exp(-sum * two_pi), 0};`, the "double + complex literal" -> "complex" conversion is a bit confusing.

File: src/icrar/leap-accelerate/algorithm/casa/PhaseRotate.cc [hunk 2]
```diff
@@ -154,53 +153,48 @@ namespace casalib
...
+                casacore::Matrix<double> cal1 = casalib::multiply(metadata.Ad1, phaseAnglesI1);
 
-                Eigen::MatrixXd e_phaseAnglesI = icrar::cpu::MatrixRangeSelect(e_phaseAngles, ToVector(metadata.I), Eigen::all);
+                Eigen::MatrixXd e_phaseAnglesI = cpu::MatrixRangeSelect(e_phaseAngles, ToVector(metadata.I), Eigen::all);
                 casacore::Matrix<double> phaseAnglesI = ConvertMatrix(e_phaseAnglesI);
 
                 // Calculate DInt
                 casacore::Matrix<double> dInt = casacore::Matrix<double>(metadata.I.size() + 1, phaseAngles.shape()[1]);
                 dInt = 0;
+                constexpr double two_pi = boost::math::constants::pi<double>();
```
Comment (rtobar) [inline src/icrar/leap-accelerate/algorithm/casa/PhaseRotate.cc:186]: This is defined twice in this compilation unit, maybe could define it at the namespace level and reuse it.

File: src/icrar/leap-accelerate/algorithm/casa/PhaseRotate.cc [hunk 3]
```diff
@@ -239,22 +233,8 @@ namespace casalib
...
-                - metadata.phase_centre_dec_rad * metadata.oldUVW[baseline](1)
-            );
-            shiftFactor -=
-            (
-                //NOTE: polar direction
-                direction.get()[0] * uvw[baseline](0)
-                - direction.get()[1] * uvw[baseline](1)
-            );
-            shiftFactor *= two_pi;
+            constexpr double two_pi = 2 * boost::math::constants::pi<double>();
```
Comment (rtobar) [inline src/icrar/leap-accelerate/algorithm/casa/PhaseRotate.cc:236]: This is the other instance of the same constant definition.

File: src/icrar/leap-accelerate/algorithm/cpu/PhaseRotate.cc [hunk 4]
```diff
@@ -156,26 +154,20 @@ namespace cpu
...
-        // PhaseAngles I
-        // Value at last index of phaseAnglesI must be 0 (which is the reference antenna phase value)
-        Eigen::MatrixXd phaseAnglesI = icrar::cpu::MatrixRangeSelect(phaseAngles, metadata.GetI(), Eigen::all);
-        phaseAnglesI.conservativeResize(phaseAnglesI.rows() + 1, phaseAnglesI.cols());
-        phaseAnglesI(phaseAnglesI.size() - 1) = 0;
-
         Eigen::VectorXd cal1 = metadata.GetAd1() * phaseAnglesI1;
         Eigen::MatrixXd dInt = Eigen::MatrixXd::Zero(metadata.GetI().size(), metadata.GetAvgData().cols());
-
+        constexpr double two_pi = boost::math::constants::pi<double>();
```
Comment (rtobar) [inline src/icrar/leap-accelerate/algorithm/cpu/PhaseRotate.cc:166]: And again the same constant... maybe makes sense to define it somewhere more centrally?

File: src/icrar/leap-accelerate/algorithm/cuda/PhaseRotate.cu [hunk 5]
```diff
@@ -195,19 +195,14 @@ namespace cuda
...
-        // PhaseAngles I
-        // Value at last index of phaseAnglesI must be 0 (which is the reference antenna phase value)
-        Eigen::MatrixXd phaseAnglesI = icrar::cpu::MatrixRangeSelect(phaseAngles, metadata.GetI(), Eigen::all);
-        phaseAnglesI.conservativeResize(phaseAnglesI.rows() + 1, phaseAnglesI.cols());
-        phaseAnglesI(phaseAnglesI.size() - 1) = 0;
-
         Eigen::VectorXd cal1 = metadata.GetAd1() * phaseAnglesI1;
         
         Eigen::MatrixXd dInt = Eigen::MatrixXd::Zero(metadata.GetI().size(), metadata.GetAvgData().cols());
+        constexpr double two_pi = boost::math::constants::pi<double>();
```
Comment (rtobar) [inline src/icrar/leap-accelerate/algorithm/cuda/PhaseRotate.cu:201]: Same constant

File: src/icrar/leap-accelerate/model/cpu/MetaData.cc [hunk 6]
```diff
@@ -189,24 +190,48 @@ namespace cpu
...
+
+        m_dd3 <<
+        1,              0,               0,
+        0, std::cos(ang3), -std::sin(ang3),
+        0, std::sin(ang3),  std::cos(ang3);
+
+        m_lmn = Eigen::Vector3d::Zero();
+
+        m_dd = m_dd3 * m_dd2;
+        LOG(info) << "m_dd check: " << m_dd;
```
Comment (rtobar) [inline src/icrar/leap-accelerate/model/cpu/MetaData.cc:224]: maybe log in `debug`?

File: src/icrar/leap-accelerate/model/cpu/MetaData.cc [hunk 7]
```diff
@@ -189,24 +190,48 @@ namespace cpu
...
+        m_dd = m_dd * m_dd1;
+        LOG(info) << "m_dd check: " << m_dd;
```
Comment (rtobar) [inline src/icrar/leap-accelerate/model/cpu/MetaData.cc:226]: ditto

**Review by rtobar — APPROVED** [review]
Looks pretty good, I didn't find anything major other than maybe a constant that could be reused.

A note for the future: I noticed that the PR contains also quite a bit of code cleanup (always welcomed!), probably triggered by clang-tidy or similar tools; that could probably have been managed separately though to help keeping PRs clean and to the point.

File: src/icrar/leap-accelerate/algorithm/casa/PhaseRotate.cc (hunk 2 above)
Comment (calgray) [reply src/icrar/leap-accelerate/algorithm/casa/PhaseRotate.cc:186]: Interestingly this is pi and not 2 pi. I'll move this to constants.h

File: src/icrar/leap-accelerate/ms/MeasurementSet.cc [hunk 8]
```diff
@@ -175,6 +176,49 @@ namespace icrar
...
+    {
+        auto nBaselines = GetNumBaselines();
+        Eigen::Matrix<bool, -1, 1> baselineFlags = Eigen::Matrix<bool, -1, 1>::Zero(nBaselines); 
+
+        // Filter short baselines
+        if(minimumBaselineThreshold > 0.0)
+        {
+            auto uvwShape = m_msmc->uvw().getColumn().shape();
+            auto uvSlice = casacore::Slicer(casacore::IPosition(2,0,0), casacore::IPosition(2,1,uvwShape[1]), casacore::IPosition(2,1,1));
+            casacore::Matrix<double> uv = m_msmc->uvw().getColumn()(uvSlice);
```
Comment (calgray) [inline src/icrar/leap-accelerate/ms/MeasurementSet.cc:190]: I suspect there is a better way of getting this

File: src/icrar/leap-accelerate/algorithm/cpu/PhaseRotate.cc [hunk 9]
```diff
@@ -147,26 +150,19 @@ namespace cpu
...
-        phaseAnglesI(phaseAnglesI.size() - 1) = 0;
-
         Eigen::VectorXd cal1 = metadata.GetAd1() * phaseAnglesI1;
         Eigen::MatrixXd dInt = Eigen::MatrixXd::Zero(metadata.GetI().size(), metadata.GetAvgData().cols());
-
         for(int n = 0; n < metadata.GetI().size(); ++n)
         {
             double sum = metadata.GetA()(n, Eigen::all) * cal1;
-            dInt(n, Eigen::all) = phaseAnglesI(n, Eigen::all).unaryExpr([=](double v) { return v - sum; });
+            dInt(n, Eigen::all) = icrar::arg(std::exp(std::complex<double>(0, -sum * two_pi<double>())) * metadata.GetAvgData()(n, Eigen::all));
```
Comment (calgray) [inline src/icrar/leap-accelerate/algorithm/cpu/PhaseRotate.cc:165]: This is based on the python code `dInt(n)=angle(GetAvgData(n)*exp(complex(0,-sum*2*pi)))`

File: src/icrar/leap-accelerate/algorithm/cpu/PhaseRotate.cc [hunk 10]
```diff
@@ -183,33 +179,18 @@ namespace cpu
...
-                metadata.GetConstants().phase_centre_ra_rad * metadata.GetOldUVW()[baseline](0)
-                - metadata.GetConstants().phase_centre_dec_rad * metadata.GetOldUVW()[baseline](1)
-            );
-            shiftFactor -=
-            (
-                polar_direction(0) * metadata.GetUVW()[baseline](0)
-                - polar_direction(1) * metadata.GetUVW()[baseline](1)
-            );
-            shiftFactor *= two_pi;
+            double shiftFactor = two_pi<double>() * (metadata.GetUVW()[baseline](2) - metadata.GetOldUVW()[baseline](2));
```
Comment (calgray) [inline src/icrar/leap-accelerate/algorithm/cpu/PhaseRotate.cc:187]: The previous shift factor accounts for offsets in the image that do not apply now

**Comment by calgray** [comment]
Sorry about the combined changes, at some point I must of merged with the branch containing small changes to DD matrix and summing in complex domain. My main concern was whether the results were incorrect due to a bug but the results of each implementation are still the same.

**Pull request merged**